### ISSUE TITLE
Any plans to make it available as a library?

### ISSUE THREAD

**Issue by shurcooL** [campoy/embedmd#10]
`embedmd` looks great and I may want to use it, but I doubt I'd use it standalone. Instead, I already have [`markdownfmt`](https://github.com/shurcooL/markdownfmt) which I run on Markdown files and I may want to incorporate it there.

But to do so, I'd need to be able to import `embedmd` as a library. All the code is currently inside the single package which is a command, that doesn't lend itself to importing.

I just wanted to ask if you had any plans to split it into a library + command? Similar to:
- https://godoc.org/golang.org/x/tools/imports - library
  - https://godoc.org/golang.org/x/tools/cmd/goimports - command
- https://godoc.org/honnef.co/go/unused - library
  - https://godoc.org/honnef.co/go/unused/cmd/unused - command
- https://godoc.org/github.com/shurcooL/markdownfmt/markdown - library
  - https://godoc.org/github.com/shurcooL/markdownfmt - command
- https://godoc.org/github.com/mholt/archiver - library
  - https://godoc.org/github.com/mholt/archiver/cmd/archiver - command

Thanks!

**Comment by campoy** [comment]
Which of the functions would you find more useful in the library?

Maybe `func process(out io.Writer, in io.Reader) error`?

**Comment by broady** [comment]
Maybe also some abstraction for filesystem/looking up filenames?

**Comment by shurcooL** [comment]
> Maybe also some abstraction for filesystem/looking up filenames?

Yes! I would definitely want a VFS abstraction for accessing the filesystem. It's better not to assume you'll have access to `os.OpenFile` directly, as that's not the case in all environments. ;) (As a bonus, it helps your tests for mocking purposes.)

I _think_ something like this could work well for your needs, since you only really need `OpenFile` method and nothing more like listing directories or doing stat (otherwise the `http.FileSystem` abstraction would be more fitting):

https://godoc.org/go/build#Context.OpenFile

However, I see that you use OpenFile for both reading **and** writing, so returning `io.ReadCloser` isn't enough. So you'll probably need to create an interface that fits. Something like https://godoc.org/golang.org/x/net/webdav#File. Except you don't need Stat, Readdir, etc.

> Maybe `func process(out io.Writer, in io.Reader) error`?

Yes, that one looks like a great start. But I would expose it with a slightly more convenient and future-proof API.

For inspiration, see similar APIs:

- https://godoc.org/golang.org/x/tools/imports#Process
- http://godoc.org/github.com/shurcooL/markdownfmt/markdown#Process
- https://godoc.org/go/parser#ParseFile
- https://godoc.org/go/format#Source

I would imagine it could look something like this:

```Go
// TODO: This is a very hard to implement interface. Probably simplify.
// You just need the ability to read the entire file, or write the entire file.
type File interface {
	io.ReadCloser
	io.WriterAt
	Truncate(int64) error
}

// Options specifies options for processing files.
type Options struct {
	// OpenFile opens named file for reading and writing.
	// If nil, then os.OpenFile is used
	func OpenFile(path string) (File, error)

	// (In the future, if you need to add some options, you can do it in a backwards compatible way.)
}

// Process ...
func Process(out io.Writer, in io.Reader, opt *Options) error {
	if opt == nil {
		// Use defaults.
		opt = &Options{}
	}
	if opt.OpenFile == nil {
		opt.OpenFile = func(name string) (File, error) { return os.OpenFile(name, os.O_RDWR, 0666) }
	}
	...
}
```

Alternatively, you could consider making it a little higher level, more like `parser.ParseFile`, and have it accept one of `filename`, `io.Reader`, or `[]byte`, but I'm not sure if this is an improvement:

```Go
// Process ...
//
// If src != nil, ParseFile parses the source from src and the filename is only used when
// recording position information. The type of the argument for the src parameter must
// be string, []byte, or io.Reader. If src == nil, ParseFile parses the file specified by filename.
func Process(filename string, src interface{}, opt *Options) (???, error)
```

I guess it's not 100% clear cut what the API would look like and it'd take some work to come up with a good one. :) But I'm happy to help more with this, or review any ideas.

**Comment by shurcooL** [comment]
Ok, I might have over-complicated my suggestions above because I didn't first look at where you need to write files. It seems you only write to the output, which would be the `out io.Writer` part of `process`. In that case, the whole thing can be simplified to just:

```Go
type Options struct {
	// ReadFile reads the file at path and returns its contents.
	// If nil, then ioutil.ReadFile is used.
	func ReadFile(path string) ([]byte, error)
}

// Process ...
func Process(out io.Writer, in io.Reader, opt *Options) error
```

The only thing left to do is to specify if the `path` parameter to Options.ReadFile is separated with slash ('/') or [`filepath.Separator`](https://godoc.org/path/filepath#Separator) (which may not be '/'). I would recommend trying to use slash as the separator, it's more general.

**Comment by shurcooL** [comment]
Woohoo, #11 has been resolved, unblocking this.

I think the API can begin to look something like this:

```Go
type Options struct {
	// ReadFile reads the file at '/'-separated path and returns its contents.
	// If nil, then path is converted to a operating system-defined file path and ioutil.ReadFile is called.
	func ReadFile(path string) ([]byte, error)
}

// Process ...
func Process(out io.Writer, in io.Reader, opt *Options) error
```

Note that the change is that to specify the `path string` parameter to `Options.ReadFile`, it is going to always be a '/'-separated path.

Thoughts?
